gem/optimise.py

```diff
@@ -413,3 +474,758 @@ def aggressive_unroll(expression):
+    # for tensors in pe_results:
+    #     array = contract(tensors, (i,), (j, k))
+    #     pe_results[tensors] = Indexed(Literal(array, '')
+    #
+    # return pe_results
+
+
+def optimise_expressions(expressions, quadrature_indices, argument_indices):
+    if propogate_failure(expressions):
+        return expressions
+    return [optimise(node, quadrature_indices, argument_indices) for node in expressions]
+
+
+def propogate_failure(expressions):
+    for n in traversal(expressions):
+        if isinstance(n, Failure):
+            return True
+    return False
+
+
+class Kernel_Optimiser(object):
+    """
+    An object holding a representation of a gem IR (as sum of products) and
+    perform optimisations which preserve the sematics of the IR.
+    Fields:
+    1. node
+    2. multiindex
+    3. arg_ind
+    4. rep: a list of dictionaries (summands) of list of factors, such that
+    node = Sum(summands), where a summand is Product(factors), possibly contracted
+    with multiindex
+    Factors are indexed with keys:
+        1. One of the argument indices (e.g. j):
+            factors depend on j but not other argument indices
+            All summands will have this item (to avoid checking existence),
+            with the values possibly as empty list
+        2. Combination of >1 argument indices (e.g. j, k):
+            factors depend on j and k, but not other argument indices
+            These are rarer and are added as factors are encourtered
+        3. string 'const':
+            factors with no free indices
+        4. string 'quad':
+            factors depend on indices (typically quadrature indices for reduction)
+            other than argument indices
+    """
+    def __init__(self, node, arg_ind, rep=None):
+        """
+        Constructor
+        :param node: gem node
+        :param arg_ind: tuples of tuples of argument (linear) indices
+        """
+        self.node = node
+        self.arg_ind = arg_ind
+        # TODO: make these two properties of the object
+        self.arg_ind_flat = tuple([i for id in self.arg_ind for i in id])
+        self.arg_ind_set = set(self.arg_ind_flat)
+        if isinstance(node, IndexSum):
+            self.multiindex = node.multiindex
+            self.node = node.children[0]
+        else:
+            self.multiindex = ()
+        if rep:
+            self.rep = rep
+        else:
+            self._build_repr()
+
+
+    def _decide_key(self, factor):
+        """
+        Helper function to decide the appropriate key of a factor
+        :param factor: gem node
+        """
+        fi = factor.free_indices
+        if not fi:
+            return 'const'
+        else:
+            ind_set = set(fi) & arg_ind_set
+            if len(ind_set) > 1:
+                return tuple(i for i in arg_ind_flat if i in ind_set)
+            elif len(ind_set) == 0:
+                return 'quad'
+            else:
+                return ind_set.pop()
```

Comment:
Why not make the single index case a 1-tuple, then you have:

```python
fi = factor.free_indices
if not fi:
    return "const"
else:
    indices = tuple(idx for idx in self.argument_indices if idx in fi)
    if indices == ():
        return "quad" # I would prefer "other" ? since it may not be a quad index
    return indices
```